src/async_traits.rs

```diff
@@ -48,40 +64,175 @@ where
+        self.run_with_shared_connection(|conn| Conn::TransactionManager::commit_transaction(conn))
+            .await?;
+        Ok(())
+    }
+
+    #[doc(hidden)]
+    async fn rollback_transaction(self: &Arc<Self>) -> Result<(), DieselError> {
+        self.run_with_shared_connection(|conn| {
+            Conn::TransactionManager::rollback_transaction(conn)
+        })
+        .await?;
+        Ok(())
+    }
+
+    /// Issues a function `f` as a transaction.
+    ///
+    /// If it fails, asynchronously calls `retry` to decide if to retry.
+    #[cfg(feature = "cockroach")]
+    async fn transaction_async_with_retry<R, Func, Fut, RetryFut, RetryFunc, 'a>(
+        &'a self,
+        f: Func,
+        retry: RetryFunc,
+    ) -> Result<R, DieselError>
+    where
+        R: Send + 'static,
+        Fut: Future<Output = Result<R, DieselError>> + Send,
+        Func: Fn(SingleConnection<Conn>) -> Fut + Send + Sync,
+        RetryFut: Future<Output = bool> + Send,
+        RetryFunc: Fn() -> RetryFut + Send + Sync,
+    {
+        // Check out a connection once, and use it for the duration of the
+        // operation.
+        let conn = Arc::new(self.get_owned_connection());
+
+        // Refer to CockroachDB's guide on advanced client-side transaction
+        // retries for the full context here. In short, they expect a particular
+        // name for this savepoint, but Diesel has Opinions on savepoint names,
+        // so we use this session variable to identify that any name is valid.
+        //
+        // TODO: It may be preferable to set this once per connection -- but
+        // that'll require more interaction with how sessions with the database
+        // are constructed.
+        Self::start_transaction(&conn).await?;
+        conn.run_with_shared_connection(|conn| {
+            conn.batch_execute("SET LOCAL force_savepoint_restart = true")
+        })
+        .await?;
+
+        loop {
+            // Add a SAVEPOINT to which we can later return.
+            Self::add_retry_savepoint(&conn).await?;
+
+            let async_conn = SingleConnection(Self::as_async_conn(&conn).0.clone());
+            match f(async_conn).await {
+                Ok(value) => {
+                    // The user-level operation succeeded: try to commit the
+                    // transaction by RELEASE-ing the retry savepoint.
+                    if let Err(err) = Self::commit_transaction(&conn).await {
+                        // We're still in the transation, but we at least
+                        // tried to ROLLBACK to our savepoint.
+                        if !retryable_error(&err) || !retry().await {
+                            // Bail: ROLLBACK the initial BEGIN statement too.
+                            let _ = Self::rollback_transaction(&conn).await;
+                            return Err(err);
+                        }
+                        // ROLLBACK happened, we want to retry.
```

Comment:
To make sure I understand, we're re-issuing the `SAVEPOINT ...` statement, which from CRDB's docs means we'll retry the transaction, but _not_ increase the priority. From [here](https://www.cockroachlabs.com/docs/stable/advanced-client-side-transaction-retries), the latter would require use to issue `ROLLBACK TO SAVEPOINT` instead. 

(1) Is that accurate?
(2) Is that what we want, or do we need any way to control which of those two retry paths is taken?

Reply:
Hrmmm so Cockroachdb isn't super transparent about prioritization under test, but I read this statement:

> 4. If a statement returns a retry error (identified via the 40001 error code or "restart transaction" string at the start of the error message), you can issue the [ROLLBACK TO SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/rollback-transaction) statement to restart the transaction and increase the transaction's priority. Alternately, the original [SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/savepoint) statement can be reissued to restart the transaction.
> You must now issue the statements in the transaction again.
> ...
>
> 5. Once the transaction executes all statements without encountering contention errors, execute [RELEASE SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/release-savepoint) to commit the changes. If this succeeds, all changes made by the transaction become visible to subsequent transactions and are guaranteed to be durable if a crash occurs.
>
> In some cases, the [RELEASE SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/release-savepoint) statement itself can fail with a retry error, mainly because transactions in CockroachDB only realize that they need to be restarted when they attempt to commit. If this happens, the retry error is handled as described in step 4.

This is, to me, case (5) -- the user operations have succeeded, and we issued `RELEASE SAVEPOINT`, and saw a retryable error.

This means we need to do the following:

> If this happens, the retry error is handled as described in step 4.

I interpreted this to mean "You should re-issue the savepoint and try again", since we already tried to RELEASE it. I believe this is what we do when we hit the `continue` statement below, since we unconditionally issue the `SAVEPOINT` at the start of the loop

Reply:
Ok, that all makes sense. We have:

- Issued `BEGIN`
- Issued `SAVEPOINT XXX`
- Successfully run some user statements
- Run `RELEASE SAVEPOINT XXX`
- Received an error on that last statement

The `continue` brings us back to `SAVEPOINT XXX`, which all makes sense.

What I was after is, do we ever want to issue `ROLLBACK TO SAVEPOINT XXX` to increase the priority? I don't have a good understanding of when to do that, and was just pointing out that this code will always run transactions at the same priority, while one might want to increase the priority for things we've had to retry.

Reply:
There are two cases to consider: User-supplied function `f` returns OK or return an error.

## `f` returns `OK`

- We call `commit_transaction` in Diesel: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#446
, which calls `RELEASE SAVEPOINT XXX`
  - If that succeeds, there is no need to retry
  - If that fails, Diesel automatically calls `ROLLBACK TO SAVEPOINT`: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#475 , and we re-issue the `SAVEPOINT` operation on the next loop

## `f` returns an error

- We call `rollback_transaction` in Diesel: https://docs.diesel.rs/master/src/diesel/connection/transaction_manager.rs.html#352 , which calls `ROLLBACK TO SAVEPOINT`
  - If that succeeds, we re-issue the `SAVEPOINT` operation on the next loop
  - If that fails, we bail out (rolling back the `BEGIN` statement too, but we stop retrying)

So it's through Diesel, but we **do** issue `ROLLBACK TO SAVEPOINT` either explicitly through `rollback_transaction`, or implicitly through `commit_transaction`.

Reply:
Ok, I read this:

> 4. If a statement returns a retry error (identified via the 40001 error code or "restart transaction" string at the start of the error message), you can issue the [ROLLBACK TO SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/rollback-transaction) statement to restart the transaction and increase the transaction's priority. Alternately, the original [SAVEPOINT](https://www.cockroachlabs.com/docs/v23.1/savepoint) statement can be reissued to restart the transaction.

to mean that those two are mutually exclusive: one can issue _either_ `ROLLBACK TO SAVEPOINT XXX` _or_ `SAVEPOINT XXX`. I think that must be a misreading, based on the links your provided to Diesel's impl and the fact that the tests here pass.

Reply:
I **think** it's okay to do both? When I try `ROLLBACK` without re-issuing the `SAVEPOINT`, Diesel's accounting for "how many layers deep we are" thinks that we are no longer within that savepoint, and throws errors later down the road (we think we're in "BEGIN + SAVEPOINT", Diesel only thinks we're in "BEGIN")

Reply:
Sounds good, thanks for walking through that!